packages/ajax/src/interceptors/cacheInterceptors.js

```diff
@@ -59,14 +59,14 @@ const isResponseSizeSupported = (responseSize, maxResponseSize) => {
 
 /**
  * Request interceptor to return relevant cached requests
- * @param {function(): string} getCacheId used to invalidate cache if identifier is changed
+ * @param {function(): Promise<string>} getCacheId used to invalidate cache if identifier is changed
  * @param {CacheOptions} globalCacheOptions
  * @returns {RequestInterceptor}
  */
 const createCacheRequestInterceptor =
   (getCacheId, globalCacheOptions) => /** @param {CacheRequest} request */ async request => {
     validateCacheOptions(request.cacheOptions);
-    const cacheSessionId = getCacheId();
+    const cacheSessionId = await getCacheId();
```

Comment:
Although `createCacheRequestInterceptor` is an async function, 
inside our code we don't use `await` anywhere (yet). This means it can be called as a sync function until now as well.

Inside out own code, we do: https://github.com/ing-bank/lion/blob/master/packages/ajax/src/interceptors/cacheInterceptors.js#L168

So inside our publicly exported `createCacheInterceptors` function we catch and return `cacheRequestInterceptor` with type `RequestInterceptor` atm. This would change it to `Promise<RequestInterceptor>`.

 So even though `createCacheRequestInterceptor` is defined as async atm, it's currently treated as a sync function in our own code (and potentially by our consumers).
 
 The best way to make it backwards compatible:
 - createCacheRequestInterceptor returns `{RequestInterceptor|Promise<RequestInterceptor>}`
 - getCacheId param is type `() => string|Promise<string>`
 
 ```suggestion
    const getCacheIdResult = getCacheId();
    const isPromise = 'then' in getCacheIdResult;
    const cacheSessionId = isPromise ? await getCacheIdResult : getCacheIdResult;
```

Ideally the full chain should be promises (so we should add await here: https://github.com/ing-bank/lion/blob/master/packages/ajax/src/interceptors/cacheInterceptors.js#L168)
. But it would be a breaking change

Reply:
So we can do the same "dual type" trick in Ajax.js where we can define the getCacheIdentifier option as string|Promise<string>.